### ISSUE TITLE
Backpropagation fails with a dtype mismatch error.

### ISSUE THREAD

**Issue by LunaticWyrm467** [huggingface/candle#1092]
### The Error
```
WithBacktrace { inner: DTypeMismatchBinaryOp { lhs: U8, rhs: F32, op: "add" }
```

### The MRP
[MRP.zip](https://github.com/huggingface/candle/files/12906840/MRP.zip)

The most important functions in this are the `forward()` and `adapt()` functions.  I've also commented a few points of interest as to what I think might be causing the problem (e.g. the masking operations).

The forward function and the adapt function are at the lines `170` and `277` respectively.

**Comment by LaurentMazare** [comment]
I think this is due to your loss being non-differentiable in some parts as it has the number of correct label etc which are discrete. I've pushed #1094 which will result in this being ignored which is the same as what PyTorch would do so you shouldn't see the error anymore but note that it's likely that your variables will not change during training because of the zero derivative.

**Comment by LunaticWyrm467** [comment]
Alright, thanks for letting me know (guess I overlooked that).
By the number of the correct label, you mean the index, right? In that case, when I append the new memories, I'll instead get the index of the newly appended memories by creating an eq mask and argmax'ing it. I'll test both this new approach and the old once the release with the fix is pushed out.

**Comment by LaurentMazare** [comment]
Closing this as it's hopefully all good now.
